### PR TITLE
router: add support for retry options predicate extensions

### PR DESCRIPTION
**Pull request by mattklein123**
This was built for Envoy Mobile, but will allow generic modification
of router behavior between retries. Currently it only supports modifying
upstream socket options (to in practice impact interface binding), but
in the future is likely to be extended to modify timeouts, retry back
off times, request headers, etc.

Risk Level: Low. New feature.
Testing: Unit tests.
Docs Changes: N/A
Release Notes: Added.
Platform Specific Features: N/A

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -1329,6 +1329,10 @@ message RetryPolicy {
   // details.
   repeated RetryHostPredicate retry_host_predicate = 5;
 
+  // Retry options predicates that will be applied prior to retrying a request. These predicates
+  // allow customizing request behavior between retries.
+  repeated core.v3.TypedExtensionConfig retry_options_predicates = 12;
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:1334]: Is there a `[#extension-category:]` that this field can be associated with?

File: source/common/router/router.cc [hunk 2]
```diff
@@ -1619,6 +1622,17 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
   return true;
 }
 
+void Filter::runRetryOptionsPredicates(UpstreamRequest& retriable_request) {
+  for (const auto& options_predicate : route_entry_->retryPolicy().retryOptionsPredicates()) {
+    Upstream::RetryOptionsPredicate::UpdateOptionsParameters parameters{
```
Comment (adisuissa) [inline source/common/router/router.cc:1627]: nit: const

**Review by adisuissa — COMMENTED** [review]
Looks good!
Left an API comment, and a small nit.

File: api/envoy/config/route/v3/route_components.proto (hunk 1 above)
Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:1334]: I think if I add this the docs will break because there are no built in extensions. Let me add a comment/TODO here to add a category if/when there are built-in extensions?

Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:1334]: Yeah confirmed this doesn't work. I will add a TODO.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: envoy/router/router.h [hunk 3]
```diff
@@ -231,6 +231,13 @@ class RetryPolicy {
    */
   virtual Upstream::RetryPrioritySharedPtr retryPriority() const PURE;
 
+  /**
+   * @return the retry options predicates for this policy. Each policy will be applied prior
+   * to retrying a request, allowing for request behavior to be customized.
+   */
+  virtual const std::vector<Upstream::RetryOptionsPredicateConstSharedPtr>&
```
Comment (ggreenway) [inline envoy/router/router.h:238]: absl::Span instead of vector?

File: envoy/upstream/retry.h [hunk 4]
```diff
@@ -92,13 +93,57 @@ class RetryHostPredicate {
...
+  struct UpdateOptionsParameters {
+    // Stream info for the previous request attempt that is about to be retried.
+    const StreamInfo::StreamInfo& retriable_request_stream_info_;
+    // The current upstream socket options that were used for connection pool selection on the
+    // previous attempt.
+    Network::Socket::OptionsSharedPtr current_upstream_socket_options_;
+  };
+
+  struct UpdateOptionsReturn {
+    // New upstream socket options to apply to the next request attempt. If changed, will effect
```
Comment (ggreenway) [inline envoy/upstream/retry.h:111]: nit: affect, not effect

File: source/common/router/router.cc [hunk 5]
```diff
@@ -1619,6 +1622,17 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
   return true;
 }
 
+void Filter::runRetryOptionsPredicates(UpstreamRequest& retriable_request) {
+  for (const auto& options_predicate : route_entry_->retryPolicy().retryOptionsPredicates()) {
+    const Upstream::RetryOptionsPredicate::UpdateOptionsParameters parameters{
```
Comment (ggreenway) [inline source/common/router/router.cc:1627]: The docs say that the parameter will be the previous attempt's options, but I think by rebuilding the options on each loop iteration, if an earlier predicate updates the options, that update will be passed to the later predicates.

File: test/common/router/router_test.cc [hunk 6]
```diff
@@ -859,13 +870,21 @@ TEST_F(RouterTest, EnvoyAttemptCountInRequestUpdatedInRetries) {
...
   // 5xx response.
   router_.retry_state_->expectHeadersRetry();
+  Upstream::RetryOptionsPredicate::UpdateOptionsReturn update_options_return{
+      std::make_shared<Network::Socket::Options>()};
+  EXPECT_CALL(*retry_options_predicate, updateOptions(_)).WillOnce(Return(update_options_return));
   Http::ResponseHeaderMapPtr response_headers1(
       new Http::TestResponseHeaderMapImpl{{":status", "503"}});
   EXPECT_CALL(cm_.thread_local_cluster_.conn_pool_.host_->outlier_detector_,
               putHttpResponseCode(503));
+  // NOLINTNEXTLINE(clang-analyzer-core.CallAndMessage)
```
Comment (ggreenway) [inline test/common/router/router_test.cc:880]: Why is this needed? This looks like a bunch of similar code throughout the file.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall. Think it's worth adding an integration test to make sure option updates really work?

/wait

File: test/common/router/router_test.cc (hunk 6 above)
Comment (mattklein123) [reply test/common/router/router_test.cc:880]: I really have no idea, but tidy was flagging this line, so I put in this exclusion.

**Comment by mattklein123** [comment]
> Think it's worth adding an integration test to make sure option updates really work?

I thought about this, but I'm not quite sure how to do this since right now it only supports setting socket options. Can you think of any socket option that I could set that would somehow be obvious in an integration test?

**Comment by ggreenway** [comment]
> > Think it's worth adding an integration test to make sure option updates really work?
> 
> I thought about this, but I'm not quite sure how to do this since right now it only supports setting socket options. Can you think of any socket option that I could set that would somehow be obvious in an integration test?

Could set a no-op option, and verify that it at least makes it use a new connpool. I don't know that there's a good one to directly verify a socket option on the remote end.

**Comment by mattklein123** [comment]
> Could set a no-op option, and verify that it at least makes it use a new connpool. I don't know that there's a good one to directly verify a socket option on the remote end. 

Do you know offhand a good no-op option? I guess I could set some TCP keep alive settings or something.

**Comment by ggreenway** [comment]
Yeah, keepalive is a good choice. Widely available, and won't do anything bad/harmful in the test.

**Comment by mattklein123** [comment]
Thanks for the nudge, it turns out that we are not setting hashKey() on most socket options, making this feature not actually work in practice. :) cc @goaway

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by ggreenway — APPROVED** [review]
LGTM
